skiko/src/awtMain/objectiveC/macos/MetalRedrawer.mm

```diff
@@ -92,12 +113,25 @@ JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_redrawer_MetalRedrawer_createMe
         /// max inflight command buffers count matches swapchain size to avoid overcommitment
         device.inflightSemaphore = dispatch_semaphore_create(device.layer.maximumDrawableCount);
 
-        if (transparency)
-        {
-            NSWindow* window = (__bridge NSWindow*) (void *) windowPtr;
+        NSWindow* window = (__bridge NSWindow*) (void *) windowPtr;
+
+        if (transparency) {
             window.hasShadow = NO;
         }
 
+        jclass redrawerClass = env->GetObjectClass(redrawer);
```

Comment:
```suggestion
        static jclass redrawerClass = env->GetObjectClass(redrawer);
```

Reply:
Are you sure this is correct in Obj-C? Will `env->GetObjectClass(redrawer);` only be called the first time? No issues with multithreading?

Also, I'm reading that `jclass` instances (but not `jmethodID`) can get garbage collected... AI says we should convert it to a global reference with `env->NewGlobalRef`.

Reply:
Not sure, we need to check.

If there are issues, we need to measure performance, or extract this code to C, or to [interop.cc](https://github.com/JetBrains/skiko/blob/master/skiko/src/jvmMain/cpp/common/interop.cc). We always used static/global variables for this in the past, so we should have a reason for doing this differently.

Reply:
> env->NewGlobalRef

Sounds right

Reply:
> No issues with multithreading?

Usually this class is created in the Swing thread

Reply:
This is called once per window, no? Doesn't sound like performance is critical.

I'll change it to a static initialized to null with an explicit check, so I don't have to store the jclass too.

Reply:
> This is called once per window, no? Doesn't sound like performance is critical.

Yes, so it can affect window startup. And if we would do this everywhere, it may be noticible every time instead of just the first time. I don't have numbers, so we either stick to the established approach, or measure.

> initialized to null with an explicit check

Why? The static code looks fine to me in the C context, we only need to check Objective C.

Reply:
> Why?

Ah, to not store jclass. Missed this ). Ok

Reply:
@igordmn updated to:
```
static jmethodID getOnOcclusionStateChangedMethodID(JNIEnv *env, jobject redrawer) {
    static jmethodID onOcclusionStateChanged = NULL;
    if (onOcclusionStateChanged == NULL) {
        jclass redrawerClass = env->GetObjectClass(redrawer);
        onOcclusionStateChanged = env->GetMethodID(redrawerClass, "onOcclusionStateChanged", "(Z)V");
    }
    return onOcclusionStateChanged;
}
```

Reply:
> Yes, so it can affect window startup. And if we would do this everywhere, it may be noticible every time instead of just the first time. I don't have numbers, so we either stick to the established approach, or measure.

Tested out of curiosity:
```
        struct timespec tstart={0,0}, tend={0,0};
        clock_gettime(CLOCK_MONOTONIC, &tstart);
        jclass redrawerClass = env->GetObjectClass(redrawer);
        onOcclusionStateChanged = env->GetMethodID(redrawerClass, "onOcclusionStateChanged", "(Z)V");
        clock_gettime(CLOCK_MONOTONIC, &tend);
        printf("Getting onOcclusionStateChanged method ID took %d ns.\n", tend.tv_nsec - tstart.tv_nsec);
```

prints 
```
Getting onOcclusionStateChanged method ID took 4000 ns.
```